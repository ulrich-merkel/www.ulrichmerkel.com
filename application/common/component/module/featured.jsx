/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.4
 *
 * @requires react
 * @requires classnames
 * @requires shortid
 * @requires common/component/module/featured/item
 *
 * @changelog
 * - 0.0.4 Excluded item into separate component
 * - 0.0.3 Moved to stateless function
 * - 0.0.2 Rewritten for es2015
 * - 0.0.1 Basic functions and structure
 */
import React, { PropTypes } from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import ModuleFeaturedItem from './featured/item';

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {Object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
function ModuleFeatured(props) {

    const {
        componentType,
        className,
        itemType,
        content,
        children,
        ...otherProps
    } = props;

    if (!content || !content.list || !content.list.length) {
        return null;
    }

    const ComponentType = componentType;
    const componentClassName = classnames(
        'm-featured',
        className
    );
    let componentSchema = {};

    if (itemType) {
        componentSchema = {
            itemScope: true,
            itemType
        };
    }

    return (
        <ComponentType
            className={componentClassName}
            role='list'
            {...componentSchema}
            {...otherProps}
        >
            {content.list.map((value) => {
                return (
                    <ModuleFeaturedItem
                        key={shortid.generate()}
                        path={value.path}
                        headline={value.headline}
                        lead={value.lead}
                        img={value.img}
                    />
                );
            })}
            {children}
        </ComponentType>
    );

}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @property {string} [componentType='ul'] - The component element type used for React.createElement
 * @property {string} [className] - The component css class names - will be merged into component default classNames
 * @property {string} [itemType='https://schema.org/ItemList'] - The schema.org itemtype url attribute
 * @property {Array|string} [children] - The component dom node childs - usally an array of components, if there is only a single child it's a string
 * @property {Object} [content='{}'] - The component translation config
 */
ModuleFeatured.propTypes = {
    componentType: PropTypes.string,
    className: PropTypes.string, // eslint-disable-line react/require-default-props
    itemType: PropTypes.string,
    children: PropTypes.node, // eslint-disable-line react/require-default-props
    content: PropTypes.shape({
        list: PropTypes.arrayOf(
            PropTypes.shape({
                path: PropTypes.string,
                headline: PropTypes.string,
                lead: PropTypes.string,
                img: PropTypes.object
            })
        )
    })
};

/**
* Set defaults if props aren't available.
*
* @static
* @type {Object}
* @see ModuleFeatured.propTypes
*/
ModuleFeatured.defaultProps = {
    componentType: 'ul',
    itemType: 'https://schema.org/ItemList',
    content: {}
};

export default ModuleFeatured;
