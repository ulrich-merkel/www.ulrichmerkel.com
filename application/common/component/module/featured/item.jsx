/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.4
 *
 * @requires React
 * @requires React-Router
 * @requires classnames
 * @requires component/element/image
 *
 * @see {@link http://stackoverflow.com/questions/30115324/pass-props-in-link-react-router}
 *
 * @changelog
 * - 0.0.4 excluded headline/lead into separate component
 * - 0.0.3 moved to stateless function
 * - 0.0.2 rewritten for es2015
 * - 0.0.1 basic functions and structure
 *
 * @example <caption>Example usage (jsx)</caption>
 * import Featured from './featured';
 */
import React, { PropTypes } from 'react';
import { Link } from 'react-router';
import classnames from 'classnames';

import Picture from './../../element/picture';

/**
 * Function representing a component to return a single react child element.
 *
 * This React component is defined as a plain JavaScript function.
 * In an ideal world, most of the components would be stateless functions,
 * because in the future weâ€™ll also be able to make performance optimizations
 * specific to these components by avoiding unnecessary checks and memory allocations.
 * This is the recommended pattern, when possible.
 *
 * @constructor
 * @private
 * @param {Object} [props] The current component props
 * @returns {React.Element} React component markup
 */
function ModuleFeaturedItem(props) {

    const {
        path,
        headline,
        img,
        children
    } = props;

    const componentListItemClassName = classnames('m-featured__item');

    return (
        <Link
            to={`work/${path}`}
            className={componentListItemClassName}
            title={headline}
            itemProp='itemListElement'
            itemScope
            itemType='http://schema.org/SiteNavigationElement'
        >
            <span className='m-featured__hover'>
                <span className='m-featured__text'>
                    <span aria-level='3' className='m-featured__headline' role='heading'>
                        {headline}
                    </span>
                    <span aria-hidden='true' className='m-featured__icon c-font-icon--plus' />
                </span>
            </span>

            <Picture
                name={img.name}
                ext={img.ext}
                path={img.path}
                alt={img.alt}
                sizes={img.sizes}
                className={'m-featured__image'}
            />

            {children}

            <meta itemProp='name' content={headline} />
        </Link>
    );

}

/**
 * Valiate props via React.PropTypes helpers.

 * @static
 * @type {React.Component.PropTypes}
 * @property {object} propTypes
 * @property {string} [path] The react-router link
 * @property {string} [headline] The items title content
 * @property {Object} [img] The items image config
 * @property {Array|string} [children] The component dom node childs - usally an array of components, if there is only a single child it's a string
 */
ModuleFeaturedItem.propTypes = {
    path: PropTypes.string,
    headline: PropTypes.string,
    img: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    children: PropTypes.node
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {React.Component.DefaultProps}
 * @property {Object} img='{}' The items image config
 */
ModuleFeaturedItem.defaultProps = {
    img: {}
};

export default ModuleFeaturedItem;
