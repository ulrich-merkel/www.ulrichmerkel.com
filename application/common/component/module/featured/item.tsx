/**
 * Es6 module for React Component.
 * Component module React classes combine elements to
 * bigger parts of the page.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://stackoverflow.com/questions/30115324/pass-props-in-link-react-router}
 */
import { default as React, FunctionComponent, ReactNode } from 'react';
import classnames from 'classnames';

import { A } from '../../element/a';
import { Icon } from '../../element/icon';
import { Picture } from '../../element/picture';
import { Meta } from '../../element/meta';

type Props = {
    path: string;
    headline: string;
    img: {
        name: string;
        ext: "" | "jpg" | "png";
        path: string;
        alt: string;
        sizes: [];
    }
    children: ReactNode;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleFeaturedItem: FunctionComponent<Props> = (props) => {
    const { path = '', headline, img, children } = props;

    const componentListItemClassName = classnames('m-featured__item');

    return (
        <li
            className="m-featured__list-item"
            itemProp="itemListElement"
            itemScope
            itemType="http://schema.org/SiteNavigationElement"
        >
            <A
                to={`/work/${path}`}
                className={componentListItemClassName}
                title={headline}
            >
                <span className="m-featured__hover">
                    <span className="m-featured__text">
                        <span
                            aria-level={3}
                            className="m-featured__headline"
                            role="heading"
                        >
                            {headline}
                        </span>
                        <Icon className="m-featured__icon" icon="plus" />
                    </span>
                </span>

                <Picture
                    name={img.name}
                    ext={img.ext}
                    path={img.path}
                    alt={img.alt}
                    sizes={img.sizes}
                    className={'m-featured__picture'}
                />

                {children}

                <Meta itemProp="name" content={headline} />
            </A>
        </li>
    );
}
