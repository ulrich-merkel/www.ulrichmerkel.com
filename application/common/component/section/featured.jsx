/**
 * Es6 module for React Component.
 * Section components combine modules and elements
 * to create a page section which could be easily
 * used in page components.
 *
 * @file
 * @module
 * @flow weak
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.4
 *
 * @requires react
 * @requires common/component/section/common/grid
 * @requires common/component/module/article
 * @requires common/component/module/featured
 *
 * @changelog
 * - 0.0.4 Added SectionCommonGridDefault
 * - 0.0.3 Moved to stateless function
 * - 0.0.2 Rewritten for es2015
 * - 0.0.1 Basic functions and structure
 */
import React, { PropTypes } from 'react';

import SectionCommonGrid from './common/grid';
import ModuleArticle from './../module/article';
import ModuleFeatured from './../module/featured';

/**
 * Function representing a component to return a single react child element.
 *
 * @constructor
 * @param {Object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
function SectionFeatured(props) {

    const {
        children,
        content,
        ...moduleProps
    } = props;

    return (
        <SectionCommonGrid>
            <ModuleArticle {...{ content }} isSpaced noMargin>
                <ModuleFeatured {...{ content }} {...moduleProps}>
                    {children}
                </ModuleFeatured>
            </ModuleArticle>
        </SectionCommonGrid>
    );

}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {Object}
 * @property {Array|string} [children] - The component dom node childs, usally an array of components, if there is only a single child it's a string
 * @property {Object} [content] - The content config input
 */
SectionFeatured.propTypes = {
    children: PropTypes.node,
    content: PropTypes.object
};

export default SectionFeatured;
