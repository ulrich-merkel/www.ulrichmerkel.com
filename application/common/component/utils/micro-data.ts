/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
