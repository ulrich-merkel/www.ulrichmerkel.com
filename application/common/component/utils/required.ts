/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
