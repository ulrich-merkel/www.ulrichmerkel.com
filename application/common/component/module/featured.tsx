/**
 * Es6 module for a featured module.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent, ReactNode } from 'react';
import classnames from 'classnames';
import shortid from 'shortid';

import { isValidArray } from '../../utils/array';
import { List } from '../element/list';
import { ModuleFeaturedItem } from './featured/item';

type Props = {
    children?: ReactNode;
    className?: string;
    content?: {
        list: {
            path: string;
            headline: string;
            lead: string;
            img: Record<string, unknown>;
        }[];
    };
    itemType?: string;
    role?: string;
};

/**
 * Function representing a featured module.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleFeatured: FunctionComponent<Props> = (props) => {
    const { children, className, content, itemType, role } = props;

    if (!isValidArray(content?.list)) {
        return null;
    }

    const componentClassName = classnames('m-featured', className);

    return (
        <List className={componentClassName} {...{ itemType, role }}>
            {content.list.map(function fnMap(value) {
                return (
                    <ModuleFeaturedItem
                        key={shortid.generate()}
                        path={value.path}
                        headline={value.headline}
                        lead={value.lead}
                        img={value.img}
                    />
                );
            })}
            {children}
        </List>
    );
};
