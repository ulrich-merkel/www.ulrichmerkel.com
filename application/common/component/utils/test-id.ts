/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
