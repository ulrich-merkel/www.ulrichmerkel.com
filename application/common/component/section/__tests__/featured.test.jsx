import * as React from 'react';
import { render } from '../../../../__tests__/utils/test-utils';
import { SectionFeatured } from '../featured';

describe('SectionFeatured', function fnDescribe() {
    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <SectionFeatured content={{}}>
                Section featured children
            </SectionFeatured>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
