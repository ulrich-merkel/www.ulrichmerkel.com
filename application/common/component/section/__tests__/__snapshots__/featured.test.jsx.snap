// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionFeatured should render correctly 1`] = `
<DocumentFragment>
  <section
    class="l-grid l-main__transition"
  >
    <div
      class="l-grid__row"
    >
      <div
        class="l-grid__col--12"
      >
        <article
          class="has-no-margin m-article"
          itemscope=""
          itemtype="https://schema.org/Article"
        >
          <div
            class="m-article__text"
            itemprop="text"
          />
        </article>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
